src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
+
+		log_info("Started pg_autoctl %s service with pid %d",
+				 service->name, service->pid);
+	}
+	else
+	{
+		log_error("Failed to start pg_autoctl %s service",
+				  service->name);
+		(void) unblock_signals(&sig_mask_orig);
+		return false;
+	}
+
+	/*
+	 * IGNORE-BANNED fits here because we want to take ownership of the service.
+	 * We copy to an address that we own from a non overlapping address.
+	 */
+	dynamicService = &supervisor->dynamicServices.services[serviceIndex];
+	memcpy(dynamicService, service, sizeof(*service)); /* IGNORE-BANNED */
+	supervisor->dynamicServices.serviceCount++;
+
+	(void) unblock_signals(&sig_mask_orig);
+
+	return true;
+}
+
+
+/*
+ * supervisor_remove_dynamic stops a running dynamic service with serviceName.
+ * Once the service is stopped, then it is removed from the dynamicService array
+ * and supervisor_loop can largely forget about it.
+ *
+ * However, it will get notified during waitpid, so we add this service to the
+ * recentlyRemoved array and we will hold it there until it is requested via
+ * supervisor_find_recently_removed_dynamic() which subsequently removes it.
+ *
+ * Returns true if the service is stopped and removed from the dynamicService
+ * array, false otherwise.
+ */
+bool
+supervisor_remove_dynamic(Supervisor *supervisor, const char *serviceName)
+{
+	Service *dynamicService = NULL;
+	int serviceCount = supervisor->dynamicServices.serviceCount;
+	int serviceIndex;
+
+	/* Find it in the dynamic array */
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &supervisor->dynamicServices.services[serviceIndex];
+
+		if (!strncmp(service->name, serviceName, strlen(serviceName)))
+		{
+			dynamicService = service;
+			break;
+		}
+	}
+
+	/* It was not found, nothing to do. */
+	if (!dynamicService)
+	{
+		return false;
+	}
+
+	/*
+	 * Since we are going to signal the service, we have to make certain that we
+	 * will not error out in the supervisor_loop on the waitpid() response. So
+	 * after killing the service and removing it from the dynamicService array,
+	 * we add it to the recentlyRemoved array to find it again later.
+	 */
+	if (!kill(dynamicService->pid, SIGTERM))
+	{
+		log_info("Service %s with %d did not receive signal",
+				 dynamicService->name, dynamicService->pid);
+	}
+
+	log_info("Stopped pg_autoctl %s service with pid %d",
+			 dynamicService->name, dynamicService->pid);
+
+	/*
+	 * Remove the dynamic Service from the array by swapping it out with the
+	 * last element in the array and bringing the array count down by one
+	 */
+	supervisor->dynamicServices.services[serviceIndex] =
+		supervisor->dynamicServices.services[serviceCount - 1];
+	supervisor->dynamicServices.serviceCount--;
+
+	/*
+	 * Finally add the dynamic service to the recently removed array.
+	 *
+	 * If we reached the end of this array, start again at the beginning. This
+	 * is a loosely held array and we do not fuss too much.
+	 */
```

Comment:
Oh, then I finally get what this array is useful for. What about renaming it to `pendingShutdown` or something that expresses to the reader that we keep track of services we signaled for termination so that we can match them later when `waitpid()` confirms their termination?

Reply:
Rights, it is not very obvious at the beginning. I will rename